v3/src/components/data-display/components/legend/categorical-legend-model.ts

```diff
@@ -0,0 +1,254 @@
+
+export interface Key {
+  category: string;
+  color: string;
+  index: number;
+  column: number;
+  row: number;
+}
+interface Layout {
+  maxWidth: number;
+  fullWidth: number;
+  numColumns: number;
+  numRows: number;
+  columnWidth: number;
+}
+
+export const keySize = 15
+export const padding = 5
+export const labelHeight = getStringBounds('Wy', vars.labelFont).height + axisGap
+
+export class CategoricalLegendModel {
+  dragInfo = {
+    category: "",
+    initialOffset: { x: 0, y: 0 },
+    currentDragPosition: { x: 0, y: 0 },
+    initialIndex: -1,
+    currentIndex: -1
+  }
+
+  dataDisplayLayout: DataDisplayLayout
+  dataConfiguration: IDataConfigurationModel | undefined
+
+  categoriesAtDragStart: string[] = []
+  colorsAtDragStart: Record<string, string> = {}
+
+  constructor(dataConfiguration: IDataConfigurationModel | undefined, dataDisplayLayout: DataDisplayLayout) {
+    this.dataDisplayLayout = dataDisplayLayout
+    this.dataConfiguration = dataConfiguration
+    makeAutoObservable(this)
+  }
+
+  setDataConfiguration(dataConfiguration: IDataConfigurationModel | undefined) {
+    this.dataConfiguration = dataConfiguration
+  }
+
+  setDataDisplayLayout(dataDisplayLayout: DataDisplayLayout) {
+    this.dataDisplayLayout = dataDisplayLayout
+  }
+
+  get categoriesFromDoc() {
+    const categories = this.dataConfiguration?.categoryArrayForAttrRole('legend') || []
+    return categories.filter(cat => cat !== kMain)
+  }
+
+  getColorForCategoryFromDoc(cat: string, defaultColor = missingColor) {
+    return this.dataConfiguration?.getLegendColorForCategory(cat) ?? defaultColor
+  }
+
+  get numCategories() {
+    return this.categoriesFromDoc.length
+  }
+
+  get categoryTextMaxWidth() {
+    let maxWidth = 0
+    this.categoriesFromDoc.forEach(cat => {
+      maxWidth = Math.max(maxWidth, measureText(cat, kDataDisplayFont))
+    })
+    return maxWidth
+  }
+
+  get layoutData() {
+    const fullWidth = this.dataDisplayLayout.tileWidth
+    const maxWidth = this.categoryTextMaxWidth + keySize + padding
+    const numColumns = Math.max(Math.floor(fullWidth / maxWidth), 1)
+    const columnWidth = fullWidth / numColumns
+    const numRows = Math.ceil((this.numCategories ?? 0) / numColumns)
+
+    const lod: Layout = {
+      fullWidth,
+      maxWidth,
+      numColumns,
+      columnWidth,
+      numRows
+    }
+
+    return lod
+  }
+
+  coordinatesToCatIndex(localPoint: { x: number; y: number; }) {
+    const lod = this.layoutData
+    const { x, y } = localPoint
+    const col = Math.floor(x / lod.columnWidth)
+    const row = Math.floor(y / (keySize + padding))
+    const catIndex = row * lod.numColumns + col
+
+    return catIndex >= 0 && catIndex < this.numCategories ? catIndex : -1
+  }
+
+  catLocation(categoryKey: Key) {
+    return {
+      x: axisGap + categoryKey.column * this.layoutData.columnWidth,
+      y: /*labelHeight + */ categoryKey.row * (keySize + padding)
```

Comment:
Do you want to keep this comment?

Reply:
It was there before. I guess since it is rendering correctly it makes sense to remove it.